File: 54-spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        int row = matrix.length;
        int col = matrix[0].length;

        int left = 0;
        int right = col - 1;
        int top = 0;
        int bottom = row - 1;

        while (ans.size() < row * col) {
            for (int i = left; i <= right && ans.size() < row * col; i++) {
                ans.add(matrix[top][i]);
            }
            top++;

            for (int i = top; i <= bottom && ans.size() < row * col; i++) {
                ans.add(matrix[i][right]);
            }
            right--;
            for (int i = right; i >= left && ans.size() < row * col; i--) {
                ans.add(matrix[bottom][i]);
            }
            bottom--;

            for (int i = bottom; i >= top && ans.size() < row * col; i--) {
                ans.add(matrix[i][left]);
            }
            left++;
        }

        return ans;
    }
}

File: 54-spiral-matrix/spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if (matrix.empty()) return {};
        
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int> result;
        
        int top = 0, bottom = m - 1;
        int left = 0, right = n - 1;
        
        while (top <= bottom && left <= right) {
            // 1. Move Right: Traverse the top row
            for (int j = left; j <= right; j++) {
                result.push_back(matrix[top][j]);
            }
            top++;
            
            // 2. Move Down: Traverse the rightmost column
            for (int i = top; i <= bottom; i++) {
                result.push_back(matrix[i][right]);
            }
            right--;
            
            // 3. Move Left: Traverse the bottom row (if it still exists)
            if (top <= bottom) {
                for (int j = right; j >= left; j--) {
                    result.push_back(matrix[bottom][j]);
                }
                bottom--;
            }
            
            // 4. Move Up: Traverse the leftmost column (if it still exists)
            if (left <= right) {
                for (int i = bottom; i >= top; i--) {
                    result.push_back(matrix[i][left]);
                }
                left++;
            }
        }
        
        return result;
    }
};